### PR TITLE
FIX do not append trailing slash if query params are present in the url

### PR DESCRIPTION
**Pull request by Dominik2809**
### All changes meet the following requirements
- [x] Changelog entry was added
- [x] Changes have been tested by the author
- [x] Changes have been tested by the reviewer
- [x] Plugin can be built

@plentymarkets/ceres-io

### CODE REVIEW

File: src/Services/UrlBuilder/UrlQuery.php [hunk 1]
```diff
@@ -86,19 +86,28 @@ public function toRelativeUrl( bool $includeLanguage = false )
         {
             return null;
         }
+    
+        $queryParams = '';
+        if($this->hasQueryParams($this->path))
```
Comment (felixdausch) [inline src/Services/UrlBuilder/UrlQuery.php]: Ist hier die Prüfung wirklich notwendig? Kannst du nicht pauschal am `?` trennen. Ggf. ist $queryParams dann nicht gefüllt.

File: src/Services/UrlBuilder/UrlQuery.php [hunk 2]
```diff
@@ -86,19 +86,28 @@ public function toRelativeUrl( bool $includeLanguage = false )
...
+        {
+            $e = explode('?', $this->path);
+        
+            $this->path = $e[0];
+            $queryParams = '?'.$e[1];
+        }
 
-        $trailingSlash = self::shouldAppendTrailingSlash() ? "/" : "";
+        $trailingSlash = self::shouldAppendTrailingSlash() && !$this->hasTralingSlash($this->path) ? "/" : "";
```
Comment (felixdausch) [inline src/Services/UrlBuilder/UrlQuery.php]: Du solltest nach dem Splitten an `?` den Trailing Slash pauschal entfernen, falls vorhanden und ihn dann ggf. wieder hinzufügen. So fängst du auch den Fall, dass ein Slash da ist aber nicht da sein soll ab.

**Review by felixdausch — CHANGES_REQUESTED** [review]

File: src/Services/UrlBuilder/UrlQuery.php [hunk 3]
```diff
@@ -86,19 +86,33 @@ public function toRelativeUrl( bool $includeLanguage = false )
...
-            return '/' . $this->lang . $this->path . $trailingSlash;
+            return '/' . $this->lang . $this->path . $trailingSlash . $queryParams;
         }
         elseif(strlen($this->path) == 0)
         {
             return '/';
         }
 
-        return $this->path . $trailingSlash;
+        return $this->path . $trailingSlash . $queryParams;
```
Comment (felixdausch) [inline src/Services/UrlBuilder/UrlQuery.php]:
```
list($path, $queryParams) = explode('?', $this->path);
if($path[strlen($path)-1] == '/')
{
    $path = substr($path, 0, -1);
}
$queryParams = strlen($queryParams) ? "?" . $queryParams : "";

$trailingSlash = self::shouldAppendTrailingSlash() ? "/" : "";

if ( $includeLanguage && strpos($path, '/'.$this->lang) !== 0)
{
    return '/' . $this->lang . $path . $trailingSlash . $queryParams;
}
elseif(strlen($path) == 0)
{
    return  "/";
}

return $path . $trailingSlash . $queryParams
```

**Review by felixdausch — CHANGES_REQUESTED** [review]

**Review by felixdausch — DISMISSED** [review]

**Review by felixdausch — APPROVED** [review]

**Comment by maxhelligeplenty** [comment]
Hallo  ,
### All changes meet the following requirements
- [ ] Changelog entry was added
- [ ] Changes have been tested by the author
- [ ] Changes have been tested by the reviewer
- [ ] Plugin can be built

@plentymarkets/ceres-io 
You can view, comment on, or merge this pull request online at:

  https://github.com/plentymarkets/plugin-io/pull/657

-- Commit Summary --

  * FIX do not append trailing slash if query params are present in the url

-- File Changes --

    M src/Services/UrlBuilder/UrlQuery.php (14)

-- Patch Links --

https://github.com/plentymarkets/plugin-io/pull/657.patch
https://github.com/plentymarkets/plugin-io/pull/657.diff

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/plentymarkets/plugin-io/pull/657


Test.test.test

**Comment by maxhelligeplenty** [comment]
Hallo  ,
felixdausch requested changes on this pull request.



> @@ -86,19 +86,28 @@ public function toRelativeUrl( bool $includeLanguage = false )
         {
             return null;
         }
+    
+        $queryParams = '';
+        if($this->hasQueryParams($this->path))

Ist hier die Prüfung wirklich notwendig? Kannst du nicht pauschal am `?` trennen. Ggf. ist $queryParams dann nicht gefüllt.

>  
-        $trailingSlash = self::shouldAppendTrailingSlash() ? "/" : "";
+        $trailingSlash = self::shouldAppendTrailingSlash() && !$this->hasTralingSlash($this->path) ? "/" : "";

Du solltest nach dem Splitten an `?` den Trailing Slash pauschal entfernen, falls vorhanden und ihn dann ggf. wieder hinzufügen. So fängst du auch den Fall, dass ein Slash da ist aber nicht da sein soll ab.

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/plentymarkets/plugin-io/pull/657#pullrequestreview-311095842

Test.test.test

**Comment by maxhelligeplenty** [comment]
Hallo  ,
felixdausch requested changes on this pull request.



>          }
         elseif(strlen($this->path) == 0)
         {
             return '/';
         }
 
-        return $this->path . $trailingSlash;
+        return $this->path . $trailingSlash . $queryParams;

```
list($path, $queryParams) = explode('?', $this->path);
if($path[strlen($path)-1] == '/')
{
    $path = substr($path, 0, -1);
}
$queryParams = strlen($queryParams) ? "?" . $queryParams : "";

$trailingSlash = self::shouldAppendTrailingSlash() ? "/" : "";

if ( $includeLanguage && strpos($path, '/'.$this->lang) !== 0)
{
    return '/' . $this->lang . $path . $trailingSlash . $queryParams;
}
elseif(strlen($path) == 0)
{
    return  "/";
}

return $path . $trailingSlash . $queryParams
```

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/plentymarkets/plugin-io/pull/657#pullrequestreview-311134699

Test.test.test

**Comment by maxhelligeplenty** [comment]
Hallo  ,
felixdausch approved this pull request.





-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/plentymarkets/plugin-io/pull/657#pullrequestreview-311142945

Test.test.test

**Pull request merged**